cluster/photon-controller/util.sh

```diff
@@ -0,0 +1,995 @@
+            fi
+        done
+    done
+}
+
+#
+# Copy the certificate/key from the Kubernetes master
+# These are used to create the kubeconfig file, which allows
+# users to use kubectl easily
+#
+function copy-kube-certs {
+    local cert="kubecfg.crt"
+    local key="kubecfg.key"
+    local ca="ca.crt"
+    local cert_dir="/srv/kubernetes"
+
+    kube::log::status "Copying credentials from ${KUBE_MASTER}"
+
+    KUBE_CERT="${KUBE_TEMP}/${cert}"
+    KUBE_KEY="${KUBE_TEMP}/${key}"
+    CA_CERT="${KUBE_TEMP}/${ca}"
+    CONTEXT="photon-${INSTANCE_PREFIX}"
+
+    run-ssh-cmd ${KUBE_MASTER_IP} "sudo chmod 644 ${cert_dir}/${cert}"
+    run-ssh-cmd ${KUBE_MASTER_IP} "sudo chmod 644 ${cert_dir}/${key}"
+    run-ssh-cmd ${KUBE_MASTER_IP} "sudo chmod 644 ${cert_dir}/${ca}"
+
+    copy-file-from-vm ${KUBE_MASTER_IP} ${cert_dir}/${cert} ${KUBE_CERT}
+    copy-file-from-vm ${KUBE_MASTER_IP} ${cert_dir}/${key}  ${KUBE_KEY}
+    copy-file-from-vm ${KUBE_MASTER_IP} ${cert_dir}/${ca}   ${CA_CERT}
+
+    run-ssh-cmd ${KUBE_MASTER_IP} "sudo chmod 600 ${cert_dir}/${cert}"
+    run-ssh-cmd ${KUBE_MASTER_IP} "sudo chmod 600 ${cert_dir}/${key}"
+    run-ssh-cmd ${KUBE_MASTER_IP} "sudo chmod 600 ${cert_dir}/${ca}"
+}
+
+#
+# Copies a script to a VM and runs it
+# Parameters:
+#   - IP of VM
+#   - Path to local file
+#
+function run-script-remotely {
+    local vm_ip=$1
+    local local_file="$2"
+    local base_file=$(basename ${local_file})
+    local remote_file="/tmp/$base_file"
+
+    copy-file-to-vm ${vm_ip} ${local_file} ${remote_file}
+    run-ssh-cmd ${vm_ip} "chmod 700 ${remote_file}"
+    run-ssh-cmd ${vm_ip} "nohup sudo ${remote_file} < /dev/null 1> ${remote_file}.out 2>&1 &"
+}
+
+#
+# Runs an command on a VM using ssh
+# Parameters:
+#   - (optional) -i to ignore failure
+#   - IP address of the VM
+#   - Command to run
+# Assumes environment variables:
+#   - VM_USER
+#   - SSH_OPTS
+#
+function run-ssh-cmd {
+    local ignore_failure=""
+    if [ "$1" = "-i" ]; then
+        ignore_failure="-i"
+        shift
+    fi
+
+    local vm_ip=$1
+    shift
+
+    run-cmd ${ignore_failure} "ssh $SSH_OPTS $VM_USER@${vm_ip} $@"
+}
+
+#
+# Uses scp to copy file to VM
+# Parameters:
+#   - IP address of the VM
+#   - Path to local file
+#   - Path to remote file
+# Assumes environment variables:
+#   - VM_USER
+#   - SSH_OPTS
+#
+function copy-file-to-vm {
+    local vm_ip=$1
+    local local_file=$2
+    local remote_file=$3
+
+    run-cmd "scp $SSH_OPTS $local_file $VM_USER@${vm_ip}:$remote_file"
+}
+
+function copy-file-from-vm {
+    local vm_ip=$1
+    local remote_file=$2
+    local local_file=$3
+
+    run-cmd "scp $SSH_OPTS $VM_USER@${vm_ip}:$remote_file $local_file"
+}
+
+#
+# Run a command, print nice error output
+# Used by copy-file-to-vm and run-ssh-cmd
+#
+function run-cmd {
+    local rc=0
+    local ignore_failure=""
+    if [ "$1" = "-i" ]; then
+        ignore_failure=$1
+        shift
+    fi
+
+    local cmd=$@
```

Comment:
Assigning an array to a string! Assign as array, or use * instead of @ to concatenate.